File: src/tasks/cli/command/graph/GraphParser.java
package tasks.cli.command.graph;

import static tasks.cli.parser.ParserUtil.assertNoExtraArgs;

import org.apache.commons.cli.CommandLine;
import tasks.cli.parser.CommandParser;
import tasks.cli.parser.ParserUtil;

/** Parser for graph/xl command. */
public final class GraphParser implements CommandParser<GraphArguments> {

  @Override
  public GraphArguments parse(CommandLine commandLine) {
    /*
     * No non-flag parameters allowed
     * optional --completed flag
     * optional --all flag
     */
    assertNoExtraArgs(commandLine);

    boolean isAllSet = ParserUtil.getFlagPresence(commandLine, GraphCommand.ALL_OPTION.value());
    boolean isCompletedSet =
        isAllSet || ParserUtil.getFlagPresence(commandLine, GraphCommand.COMPLETED_OPTION.value());
    boolean isUncompletedSet = isAllSet || !isCompletedSet;

    return new GraphArguments(isCompletedSet, isUncompletedSet);
  }
}

File: src/tasks/cli/feature/graph/GraphParser.kt
package tasks.cli.feature.graph

import omnia.data.cache.Memoized
import omnia.data.structure.List
import org.apache.commons.cli.CommandLine
import tasks.cli.feature.blockers.BlockersCommand
import tasks.cli.parser.CommandParser
import tasks.cli.parser.ParseResult
import tasks.cli.parser.Parser
import tasks.cli.parser.ParserUtil
import tasks.model.Task

/** Parser for graph/xl command.  */
class GraphParser(
    private val taskParser: Memoized<out Parser<out List<out ParseResult<out Task>>>>)
    : CommandParser<GraphArguments> {
  override fun parse(commandLine: CommandLine): GraphArguments {
    /*
     * No non-flag parameters allowed
     * optional --all flag
     * optional --related tasks
     */
    ParserUtil.assertNoExtraArgs(commandLine)
    val isAllSet = ParserUtil.getFlagPresence(commandLine, GraphCommand.ALL_OPTION.value())
    val tasksToRelateTo = taskParser.value().parse(
        ParserUtil.getOptionValues(commandLine, GraphCommand.RELATED_TASKS_OPTION.value()))
    return GraphArguments(
      isAllSet, ParserUtil.extractSuccessfulResultsOrThrow(tasksToRelateTo))
  }
}